File: 213_House_Robber_II.cpp
class Solution {
private:
    int worker(vector<int>& nums, int offset) {
        int n = nums.size() - (offset == 0 ? 1 : 0);
        if (n == 0) return 0;
        else if (n == 1) return nums[0];
        int pre = 0, cur = 0;
        for (int i = offset; i < n; i++) {
            int temp = pre;
            pre = cur;
            cur = max(temp + nums[i], pre);
        }
        return cur;
    }
public:
    int rob(vector<int>& nums) {
        return max(worker(nums, 0), worker(nums, 1));
    }
};

File: 213_House_Robber_II.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) < 1:
            return 0
        if len(nums) == 1:
            return nums[0]
        
        def rob(nums):
            a = b = 0
            for n in nums:
                a, b = max(b + n, a), a
            return max(a, b)
        return max(rob(nums[:-1]), rob(nums[1:]))
